machine/v1/types_nutanixprovider.go

```diff
@@ -154,6 +164,156 @@ type NutanixResourceIdentifier struct {
 	Name *string `json:"name,omitempty"`
 }
 
+// NutanixGPUIdentifierType is an enumeration of different resource identifier types for GPU entities.
+// +kubebuilder:validation:Enum:=Name;DeviceID
+type NutanixGPUIdentifierType string
+
+const (
+	// NutanixGPUIdentifierName identifies a GPU by Name.
+	NutanixGPUIdentifierName NutanixGPUIdentifierType = "Name"
+
+	// NutanixGPUIdentifierDeviceID identifies a GPU by device ID.
+	NutanixGPUIdentifierDeviceID NutanixGPUIdentifierType = "DeviceID"
+)
+
+// NutanixGPU holds the identity of a Nutanix GPU resource in the Prism Central
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'DeviceID' ?  has(self.deviceID) : !has(self.deviceID)",message="deviceID configuration is required when type is DeviceID, and forbidden otherwise"
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'Name' ?  has(self.name) : !has(self.name)",message="name configuration is required when type is Name, and forbidden otherwise"
+// +union
+type NutanixGPU struct {
+	// type is the identifier type of the GPU device.
+	// Valid values are Name and DeviceID.
+	// +unionDiscriminator
+	// +kubebuilder:validation:Required
+	Type NutanixGPUIdentifierType `json:"type"`
+
+	// deviceID is the GPU device ID with the integer value.
+	// +optional
+	// +unionMember
+	DeviceID *int32 `json:"deviceID,omitempty"`
```

Comment:
We should have validation for minimum and maximum values for this no? Even if the maximum is some arbitrary high number?

Reply:
This is the GPU PCI device ID that is from the GPU vendor. We used it here just to identify the GPU device. I don't think we can add any range check (min, max, etc.) for it.

Reply:
Oh, but these device IDs are in hex, and are not int32, are you sure you want an int32 for this field?

Reply:
Added more context to the "gpus" filed. See L81-83:
         // gpus is a list of GPU devices to attach to the machine's VM.
	// The GPU devices should already exist in Prism Central and associated with
	// one of the Prism Element's hosts and available for the VM to attach (in "UNUSED" status).
	// +listType=set
	// +optional
	GPUs []NutanixGPU `json:"gpus"`

Reply:
I think the deviceID of the GPU attached to the PE host is related to the GPU PCI device ID (hex), but appears as an integer (from the prism-api doc and the real example). Here is the one of the GPU attached to the PE host, note that: "device_id": 8757.

host (name: karbonmsp-gpu02-1, UUID: 52362f6f-fa6f-471b-9ad0-9811fbd213de)
"gpu_list": [{"status": "UNUSED", "vendor": "NVIDIA", "num_virtual_display_heads": 0, "assignable": true, "num_vgpus_allocated": 0, "pci_address": "0000:ca:00.0", "name": "Ampere 40", "frame_buffer_size_mib": 49152, "index": 0, "max_resolution": "None", "mode": "PASSTHROUGH_COMPUTE", "fraction": 0, "guest_driver_version": "None", "device_id": 8757, "max_instances_per_vm": 0}]